### PR TITLE
*: add tidb_enable_noop_functions to default disable get_lock()/release_lock()

### PR DESCRIPTION
**Pull request by cfzjywxk**
### What problem does this PR solve? 

Fixes https://github.com/pingcap/tidb/issues/10929

This adds the variable `tidb_enable_noop_functions` to control usage of NOOP funcs in tidb such as get_lock, release_lock (always return 1), since the current behavior could be dangerous for apps relying on these functions working correctly.

The default for `tidb_enable_noop_functions` is `false`, thus users must manually set session/global to `true` to have unsafe behavior.

### What is changed and how it works?

add global/session variable "tidb_enable_noop_functions" to control usage of noop functions,
default these usage will report "function xxx has only noop implementation in tidb now, use tidb_enable_noop_functions to enable these functions"". while set true, these functions can be
used like before

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

 - Has newly added system variable

Side effects

 - Breaking backward compatibility
   applications using former version of tidbs have to set "tidb_enable_noop_functions" to true to meet their needs

Related changes

 - Need to update the documentation
 - Need to be included in the release note

### CODE REVIEW

File: expression/builtin.go [hunk 1]
```diff
@@ -625,3 +625,11 @@ var funcs = map[string]functionClass{
 	ast.JSONKeys:          &jsonKeysFunctionClass{baseFunctionClass{ast.JSONKeys, 1, 2}},
 	ast.JSONLength:        &jsonLengthFunctionClass{baseFunctionClass{ast.JSONLength, 1, 2}},
 }
+
+// some functions like "get_lock" and "release_lock" currently do NOT have
+// right implementations, but may have faked ones(like with any inputs, always return 1)
+// if apps really need these "funcs" to run, we offer sys var(tidb_use_fake_funcs) to enable fake usage
+var fakedFuncs = map[string]bool{
```
Comment (zz-jason) [inline expression/builtin.go]: you can put this set to file `expression/function_traits.go`

**Comment by zz-jason** [comment]
please follow the contributing guide to refactor the PR title: https://github.com/pingcap/community/blob/master/commit-message-pr-style.md

**Comment by cfzjywxk** [comment]
@zz-jason PTAL

**Comment by morgo** [comment]
@cfzjywxk I think a better behavior instead of "function not found" would be to suggest the user can change the setting to have syntax-only support. With the current behavior I don't believe many users will discover that such a setting exists (see https://github.com/pingcap/tidb/pull/10065 for an example error).

I also prefer a name like `tidb_skip_errors_for_unsupported_function` (maybe this is a bit long, @kolbe do you have any other ideas?)

cc @pingcap/usability-team

**Comment by kolbe** [comment]
I don't like `tidb_use_fake_funcs` very much. Having abbreviations in variable names is confusing, and I think using "fake" doesn't convey the right idea. But I also think `tidb_skip_errors_for_unsupported_function` is much too long. Since these are noop functions, I suggest we call the option `tidb_noop_functions` or `tidb_enable_noop_functions` or something along those lines.

**Comment by morgo** [comment]
> `tidb_enable_noop_functions`

+1 for this name

**Comment by cfzjywxk** [comment]
@morgo PTAL

**Comment by morgo** [comment]
LGTM

File: executor/set_test.go [hunk 2]
```diff
@@ -15,6 +15,7 @@ package executor_test
 
 import (
 	"context"
+	"github.com/pingcap/tidb/expression"
```
Comment (jackysp) [inline executor/set_test.go]: move this line to line 22

Comment (cfzjywxk) [reply executor/set_test.go]: done

File: executor/set_test.go [hunk 3]
```diff
@@ -741,3 +742,45 @@ func (s *testSuite2) TestSelectGlobalVar(c *C) {
...
+	tk.MustQuery(`select get_lock("lock", 10)`).Check(testkit.Rows("1"))
+	tk.MustQuery(`select release_lock("lock")`).Check(testkit.Rows("1"))
+
+	// restore to 0
+	tk.MustExec(`set tidb_enable_noop_functions=0;`)
+	tk.MustQuery(`select @@tidb_enable_noop_functions;`).Check(testkit.Rows("0"))
+	tk.MustQuery(`select @@global.tidb_enable_noop_functions;`).Check(testkit.Rows("0"))
+
+	_, err = tk.Exec(`select get_lock('lock2', 10);`)
+	c.Assert(terror.ErrorEqual(err, expression.ErrFunctionsNoopImpl), IsTrue, Commentf("err %v", err))
```
Comment (zz-jason) [inline executor/set_test.go]: It's better to make the comment to point out which SQL is executed.

File: store/tikv/client_fail_test.go [hunk 4]
```diff
@@ -50,6 +50,7 @@ func (s *testClientSuite) TestPanicInRecvLoop(c *C) {
 	time.Sleep(time.Second)
 	c.Assert(failpoint.Disable("github.com/pingcap/tidb/store/tikv/panicInFailPendingRequests"), IsNil)
 	c.Assert(failpoint.Disable("github.com/pingcap/tidb/store/tikv/gotErrorInRecvLoop"), IsNil)
+	time.Sleep(time.Second)
```
Comment (zz-jason) [inline store/tikv/client_fail_test.go]: why adding a `sleep`?

Comment (cfzjywxk) [reply store/tikv/client_fail_test.go]: @zz-jason  new case "client_fail_test" not stable may fail running "make test", like:

FAIL: client_fail_test.go:34: testClientSuite.TestPanicInRecvLoop

client_fail_test.go:59:
    c.Assert(err, IsNil)
... value *errors.fundamental = injected error in batchRecvLoop ("injected error in batchRecvLoop")

Comment (zz-jason) [reply store/tikv/client_fail_test.go]: TiDB 3.0 may also have this problem. /cc @hicqu

**Pull request merged**